source/ota.c

```diff
@@ -1789,15 +1671,20 @@ static OtaJobParseErr_t verifyActiveJobStatus( OtaFileContext_t * pFileContext,
             LogInfo( ( "New job document ID is identical to the current job: "
                        "Updating the URL based on the new job document." ) );
 
-            if( ( otaAgent.fileContext.pUpdateUrlPath != NULL ) && ( otaAgent.fileContext.updateUrlMaxSize == 0 ) )
+            if( otaAgent.fileContext.pUpdateUrlPath != NULL )
             {
-                otaAgent.pOtaInterface->os.mem.free( otaAgent.fileContext.pUpdateUrlPath );
-                otaAgent.fileContext.pUpdateUrlPath = pFileContext->pUpdateUrlPath;
-                pFileContext->pUpdateUrlPath = NULL;
-            }
-            else
-            {
-                memcpy( otaAgent.fileContext.pUpdateUrlPath, pFileContext->pUpdateUrlPath, otaAgent.fileContext.updateUrlMaxSize );
+                if( otaAgent.fileContext.updateUrlMaxSize == 0 )
+                {
+                    /* The buffer is allocated by us, free first then update. */
+                    otaAgent.pOtaInterface->os.mem.free( otaAgent.fileContext.pUpdateUrlPath );
+                    otaAgent.fileContext.pUpdateUrlPath = pFileContext->pUpdateUrlPath;
+                    pFileContext->pUpdateUrlPath = NULL;
+                }
+                else
+                {
+                    /* The buffer is provided by user, directly copy the new url to it. */
+                    memcpy( otaAgent.fileContext.pUpdateUrlPath, pFileContext->pUpdateUrlPath, otaAgent.fileContext.updateUrlMaxSize );
```

Comment:
Somewhat unrelated to this PR, so it could be considered in a separate PR, but shouldn't we be checking that the amount of data we are trying to copy is less than the max amount we can store? I see that it won't be possible to write outside the bounds of the array, but we may not be copying the complete URL.

Reply:
destination is `otaAgent.fileContext.pUpdateUrlPath`, and size we're copying is `otaAgent.fileContext.updateUrlMaxSize`, which is the buffer size of that destination pointer. `pFileContext->pUpdateUrlPath` I believe is pointing to network buffer.